### PR TITLE
Create a custom label method for `PrepSelPrep`

### LINKED ISSUE #7159: Drawing `PrepSelPrep` shoud be less verbose
### Feature details

`PrepSelPrep`'s LCU argument is interpreted as a parameter, resulting in a verbose drawing. The feature we'd like to have is a custom label method that treats the LCU variable as a single array, which is displayed below the tape drawing like for other arrays.

### Implementation

_No response_

### How important would you say this feature is?

1: Not important. Would be nice to have.

### Additional information

```python
import numpy as np
import pennylane as qml

k = 4
num_terms = 2**k
ops = list(qml.pauli.pauli_group(k))[:num_terms]
lcu = qml.dot(np.random.random(num_terms), ops)
@qml.qnode(qml.device("default.qubit"))
def node():
    qml.PrepSelPrep(lcu, control=range(k, 2*k))
    return qml.state()
```

```
>>> print(qml.draw(node)())
0: ─╭PrepSelPrep(0.68,0.54,0.51,0.15,0.74,0.50,0.62,0.24,0.95,0.41,0.52,0.99,0.98,0.79,0.27,0.75)─┤
1: ─├PrepSelPrep(0.68,0.54,0.51,0.15,0.74,0.50,0.62,0.24,0.95,0.41,0.52,0.99,0.98,0.79,0.27,0.75)─┤
2: ─├PrepSelPrep(0.68,0.54,0.51,0.15,0.74,0.50,0.62,0.24,0.95,0.41,0.52,0.99,0.98,0.79,0.27,0.75)─┤
3: ─├PrepSelPrep(0.68,0.54,0.51,0.15,0.74,0.50,0.62,0.24,0.95,0.41,0.52,0.99,0.98,0.79,0.27,0.75)─┤
4: ─├PrepSelPrep(0.68,0.54,0.51,0.15,0.74,0.50,0.62,0.24,0.95,0.41,0.52,0.99,0.98,0.79,0.27,0.75)─┤
5: ─├PrepSelPrep(0.68,0.54,0.51,0.15,0.74,0.50,0.62,0.24,0.95,0.41,0.52,0.99,0.98,0.79,0.27,0.75)─┤
6: ─├PrepSelPrep(0.68,0.54,0.51,0.15,0.74,0.50,0.62,0.24,0.95,0.41,0.52,0.99,0.98,0.79,0.27,0.75)─┤
7: ─╰PrepSelPrep(0.68,0.54,0.51,0.15,0.74,0.50,0.62,0.24,0.95,0.41,0.52,0.99,0.98,0.79,0.27,0.75)─┤
```

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
`PrepSelPrep` prints the coefficients of the encoded LCU as parameters in its label.
That's very verbose and impractical, see #7159.

**Description of the Change:**
Implement a custom `label` method for `PrepSelPrep` that defers the coefficients to the standard matrix cache,
which is also used for all matrix-valued parameters.
Like requested in the linked issue.

**Benefits:**
More readable drawings.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
Resolves #7159

[sc-87444]

### CODE REVIEW

File: pennylane/ops/functions/assert_valid.py [hunk 1]
```diff
@@ -319,7 +319,7 @@ def _check_capture(op):
         jaxpr = jax.make_jaxpr(lambda obj: obj)(op)
         data, _ = jax.tree_util.tree_flatten(op)
         new_op = jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, *data)[0]
-        assert op == new_op
+        assert op == new_op, f"{op}\n{new_op}"
```
Comment (dwierichs) [inline pennylane/ops/functions/assert_valid.py:322]:
```suggestion
        assert op == new_op
```

File: pennylane/templates/subroutines/prepselprep.py [hunk 2]
```diff
@@ -113,6 +113,26 @@ def map_wires(self, wire_map: dict) -> "PrepSelPrep":
...
+                # matrix in cache. use known reference in label
+                str_wo_id = f"{op_label}(M{i})"
+                break
+        else:
+            # matrix not in cache. Add to cache and use in label
+            mat_num = len(cache["matrices"])
+            cache["matrices"].append(coeffs)
+            str_wo_id = f"{op_label}(M{mat_num})"
+
+        return str_wo_id if self._id is None else f'{str_wo_id[:-1]},"{self._id}")'
```
Comment (dwierichs) [inline pennylane/templates/subroutines/prepselprep.py:134]: This is just the `operation.Operation.label` method with
- self.params[0] replaced by `coeffs`
- logic that never is true removed
- mild simplification of remaining logic.

File: pennylane/templates/subroutines/prepselprep.py [hunk 3]
```diff
@@ -113,6 +113,26 @@ def map_wires(self, wire_map: dict) -> "PrepSelPrep":
...
+        for i, mat in enumerate(cache["matrices"]):
+            if shape == qml.math.shape(mat) and qml.math.allclose(coeffs, mat):
+                # matrix in cache. use known reference in label
+                str_wo_id = f"{op_label}(M{i})"
+                break
+        else:
+            # matrix not in cache. Add to cache and use in label
+            mat_num = len(cache["matrices"])
+            cache["matrices"].append(coeffs)
+            str_wo_id = f"{op_label}(M{mat_num})"
```
Comment (willjmax) [inline pennylane/templates/subroutines/prepselprep.py:132]: I think the code is fine, but the for/else block caught me off guard. The code is very easy to understand (after I figured out how for/else works), so maybe this comment is unnecessary too.

**Review by willjmax — APPROVED** [review]
Looks good :rocket: Just two very minor suggestions.

File: doc/releases/changelog-dev.md [hunk 4]
```diff
@@ -194,6 +194,10 @@
 
 <h3>Improvements 🛠</h3>
 
+* `PrepSelPrep` now has a concise custom label method that defers the LCU coefficients to the
+  matrix cache which is also used for operations with matrix parameters.
+  [(#7164)](https://github.com/PennyLaneAI/pennylane/pull/7164)
```
Comment (isaacdevlugt) [inline doc/releases/changelog-dev.md:199]:
```suggestion
* `PrepSelPrep` now has a concise representation when drawn with `qml.draw` or `qml.draw_mpl`.
  [(#7164)](https://github.com/PennyLaneAI/pennylane/pull/7164)
```

**Review by isaacdevlugt — APPROVED** [review]
Nice! I verified with this example: 

```
>>> import pennylane as qml
>>> lcu = qml.dot([0.3, -0.1], [qml.X(2), qml.Z(2)])
>>> control = [0, 1]
>>> @qml.qnode(qml.device("default.qubit"))
... def circuit(lcu, control):
...     qml.PrepSelPrep(lcu, control)
...     return qml.state()
...
>>> print(qml.draw(circuit)(lcu, control = control))
0: ─╭PrepSelPrep(M0)─┤  State
1: ─├PrepSelPrep(M0)─┤  State
2: ─╰PrepSelPrep(M0)─┤  State

M0 =
[ 0.3 -0.1]
```

**Review by willjmax — APPROVED** [review]

**Review by albi3ro — COMMENTED** [review]
I'm not actually sure if "matrices" is the right place to hamiltonian coefficients, since it's not really a matrix.

Maybe we could start up a new cache for "target ops"? `H0`, `H1`, etc. We could then use it for all operations with a more complicated target operations, like `ApproxTimeEvolutinion`, `TrotterProduct`, `QDrift`, `Qubitization`, etc.
